### PR TITLE
add 'contextConsumer'

### PR DESCRIPTION
**Pull request by 0x53A**
closes https://github.com/fable-compiler/fable-react/issues/164

~~Note: this is currently completely untested.~~

### CODE REVIEW

**Comment by 0x53A** [comment]
It currently works on the client side, but not in SSR:

```f#
let ctx = createContext("de")

let inline consume() = contextConsumer ctx (fun lang -> str lang)

// ...

let view model =

// ...

                str "This should be the default value (de):"
                consume()
                contextProvider ctx model.Language [
                    str (sprintf "This should be the model value (%s):" model.Language)
                    consume()
                    contextProvider ctx "xyz" [
                        str (sprintf "This should be the overridden value (%s):" "xyz")
                        consume()
                    ]
                    str (sprintf "This should be the model value (%s):" model.Language)
                    consume()
                ]
                str "This should be the default value (de):"
                consume()

```


Clientside:

image


SSR:

image



I think it renders the elements too early, so that my stack pushing/popping is completely useless ...

Since SSR is currently incorrect, anyway, this is not a regression, so i think I'll revert that part and just PR  the new 'contextConsumer' function.

What do you think?

**Comment by alfonsogarciacaro** [comment]
This is due to my "cheap solution" in order to support the context API in .NET without many code changes. As it's currently difficult to pass a context object through the nested rendered elements in SSR, I'm just using the default value and _hoping_ the proper one will be used on the client side when mounting React. Does it happen like this in your example?

Maybe it'd be better to fix it in SSR (your note in the comment is already helpful!) but I'm not sure what would be the easiest way to do this. Any ideas?

In any case, thanks for the PR!

File: src/Fable.React.Helpers.fs [hunk 1]
```diff
@@ -351,6 +352,17 @@ module Helpers =
         fragment [] children
 #endif
 
+    /// Consumes a context value, either from the nearest parent in the tree, or from the default value.
+    /// Important: in SSR, this will always consume the context DEFAULT value!
+    /// More info at https://reactjs.org/docs/context.html#contextconsumer
+    let inline contextConsumer (ctx: IContext<'T>) (children: 'T -> ReactElement): ReactElement = 
+#if FABLE_COMPILER 
+        ReactBindings.React.createElement(ctx?Consumer, createObj [], [!!children])
```
Comment (alfonsogarciacaro) [inline src/Fable.React.Helpers.fs]: I think we can pass `null` here instead of `createObj []`.

**Comment by 0x53A** [comment]
> 'm just using the default value and hoping the proper one will be used on the client side when mounting React. Does it happen like this in your example?

Yes, after the first client-side render all is ok.

In my use-case (localization) it doesn't matter anyway, there is only one context anyway.

> Maybe it'd be better to fix it in SSR (your note in the comment is already helpful!) but I'm not sure what would be the easiest way to do this. Any ideas?

No simple idea, the rendering would need to be delayed so that the parent can influence the child.

**Comment by alfonsogarciacaro** [comment]
Ok, if the client side fixes this let's consider it good enough for now ;) Released as 5.2.4, thank you!

**Pull request merged**